minetest/minetest
# Texture atlas for mapblocks meshes

This PR has a trial to decrease the drawcalls count in the clientmap, mainly switches count of the textures in the OpenGL state. The crucial changes which this PR implements:

- Collecting all textures from the tile layers during the media loading and drawing them into the one atlas texture.
- Using the single huge texture (atlas) for all mapblocks meshes and their submeshes.
- Improved batching the submeshes by the same material within the mapblock.
- Avoiding unnecessary setMaterial() calls.
- Allowing the atlases to be animateable (re-draw the frames of the animated tiles).

An example of the animated atlas generated by this PR in MT:
![animated_atlas](https://github.com/user-attachments/assets/c858e456-32b6-434b-a277-1407123405a2)

Note, the atlasing is not applied in the mesh nodes as an exception for this drawtype, they use own textures spaces. There are many visual artifacts with some drawtypes, mainly UV issues, also filtering and mipmapping ones.

## Known issues (mostly small ones)

- Incorrect UVs of the `flowingliquid` drawtype nodes
- Sometimes mesh nodes whose UVs are beyond 0-1 range render on themselves an odd geometry parts
- With the filtering enabled the crack textures are rendered skewed
- With the filtering + mipmapping enabled the lower mip levels still get a pixel pollution
- The atlas is black on Android (reported by @AbduSharif)

## How to test
1. See how nodes placed in the world look with various drawtypes.
2. Enable/disable filtering, mipmapping and other graphics settings.
